pkg/cloudprovider/aws/initialization.go

```diff
@@ -0,0 +1,242 @@
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/rest"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const (
+	DefaultNodeRoleName       = "KarpenterNodeRole"
+	DefaultLaunchTemplateName = "KarpenterLaunchTemplate"
+)
+
+type Initialization struct {
+	InstanceRole    *iam.Role
+	InstanceProfile *iam.InstanceProfile
+	LaunchTemplate  *ec2.LaunchTemplate
+	Cluster         *eks.Cluster
+	ZonalSubnets    map[string]*ec2.Subnet
+}
+
+func NewInitialization(EC2API ec2iface.EC2API, EKSAPI eksiface.EKSAPI, IAMAPI iamiface.IAMAPI, config *rest.Config) *Initialization {
+	zap.S().Infof("Initializing AWS Cloud Provider")
+	cluster := clusterOrDie(EKSAPI, "etarn-dev")
+	instanceRole := instanceRoleOrDie(IAMAPI, DefaultNodeRoleName)
+	ensureAWSAuthOrDie(config, instanceRole)
+	initialization := &Initialization{
+		Cluster:         cluster,
+		InstanceRole:    instanceRole,
+		InstanceProfile: nodeInstanceProfileOrDie(IAMAPI, *instanceRole.RoleName),
+		LaunchTemplate:  launchTemplateOrDie(EC2API, cluster, DefaultLaunchTemplateName, *instanceRole.RoleName),
+		ZonalSubnets:    zonalSubnetsOrDie(EC2API, cluster),
+	}
+	zap.S().Infof("Successfully initialized AWS Cloud Provider")
+	return initialization
+}
+
+func clusterOrDie(eksAPI eksiface.EKSAPI, name string) *eks.Cluster {
+	describeClusterOutput, err := eksAPI.DescribeCluster(&eks.DescribeClusterInput{
+		Name: aws.String(name),
+	})
+	log.PanicIfError(err, "Failed to discover EKS Cluster %s", name)
+	zap.S().Infof("Successfully discovered EKS Cluster, %s", name)
+	return describeClusterOutput.Cluster
+}
+
+func nodeInstanceProfileOrDie(IAMAPI iamiface.IAMAPI, name string) *iam.InstanceProfile {
+	// 1. Detect existing InstanceProfile
+	getInstanceProfileOutput, err := IAMAPI.GetInstanceProfile(&iam.GetInstanceProfileInput{
+		InstanceProfileName: aws.String(name),
+	})
+	if err == nil {
+		zap.S().Infof("Successfully detected instance profile %s", name)
+		return getInstanceProfileOutput.InstanceProfile
+	} else if aerr, ok := err.(awserr.Error); ok && aerr.Code() != iam.ErrCodeNoSuchEntityException {
+		log.PanicIfError(err, "Failed to retrieve instance profile %s", name)
+	}
+
+	// 2. Create InstanceProfile
+	createInstanceRoleOutput, err := IAMAPI.CreateInstanceProfile(&iam.CreateInstanceProfileInput{
+		InstanceProfileName: aws.String(name),
+	})
+	log.PanicIfError(err, "Failed to create instance profile %s", name)
+	zap.S().Infof("Successfully created instance profile %s", name)
+
+	// 3. Attach Role to Instance Profile
+	_, err = IAMAPI.AddRoleToInstanceProfile(&iam.AddRoleToInstanceProfileInput{
+		InstanceProfileName: aws.String(name),
+		RoleName:            aws.String(name),
+	})
+	log.PanicIfError(err, "Failed to add role %s to instance profile %s", name, name)
+	zap.S().Infof("Successfully added role %s to instance profile %s", name, name)
+	return createInstanceRoleOutput.InstanceProfile
+}
+
+func instanceRoleOrDie(IAMAPI iamiface.IAMAPI, name string) *iam.Role {
```

Comment:
nit: can we change name -> roleName

Reply:
I'm a big fan of brief, context specific names. I learned this habit from Kubernetes (i.e. every resource name is just "name") and I quite like it. There are cases where you do need to specify if it's not clear from context, but I feel it's reasonably clear here. I you feel strongly that it's ambiguous, I'll happily change it as you ask.